### ISSUE TITLE
Regression: Choose a sane filename for created tempfiles

### ISSUE THREAD

**Issue by kaste** [SublimeLinter/SublimeLinter#1085]
The reason we had to switch to random names was bc the backend became concurrent. But we can of course just use random temporary directories and put a file in it that follows SL3.

It's simple, and it's here

https://github.com/SublimeLinter/SublimeLinter/blob/4ef4d62adc5d00ea59274e7bdd4bab828c97343c/lint/util.py#L240-L271

**Comment by braver** [comment]
So, this is to supersede #1073 by making sure the filename is the same also when the linter runs on a temp file? I think the file name alone isn't enough here, I think you need to replicate the path from the project root. So, before this is really actionable I'd like to see more clarity about what this intends to solve.

**Comment by kaste** [comment]
I don't strive to implement #1073 at all. The actionable part is to give the file a reasonable name (the SL3 name) within a temp directory.

**Comment by kaste** [comment]
@sebastienbarre Want to tackle this one?

**Comment by braver** [comment]
Ok cool. So that will support things like file-name based ignore rules etc.

**Comment by sebastienbarre** [comment]
So I see two cases.

Below, "pathname" is the full path, "[dirname](https://en.wikipedia.org/wiki/Dirname)" is "pathname" where any suffix beginning with the last slash ('/') character has been deleted, and "[basename](https://en.wikipedia.org/wiki/Basename)" is the "pathname" where any prefix up to the last slash ('/') character is deleted.

I'm not a Python guy so I assume [tempfile](https://docs.python.org/2/library/tempfile.html) can be  used to generate temporary files and directories?

## If you are not within a project

Use a temporary file, where the suffix is an underscore and the basename of the file being linted.

See: `tempfile.mkstemp([suffix=''[, prefix='tmp'[, dir=None[, text=False]]]])`, where `suffix` is underscore + basename?

Example, when linting `Payroll.php`: `/var/folders/5q/5qTPn6xq2RaWqk+1Ytw3-U+++TI/-Tmp-/tmpG7V1Y0_Payroll.php`

This will help people who lint any file they are editing (i.e. they have a global linter config). This will help linters that have rules applied based on regular expressions matching the basename. This will not help if they have rules based on pathnames, but I assume that's irrelevant here because you are not within a project.

## If you are  within a project

Use a temporary directory, and re-create the relative path to file being linted (relative to the project folder).

See: `tempfile.mkdtemp([suffix=''[, prefix='tmp'[, dir=None]]])` and [`makedirs`](https://docs.python.org/2/library/os.html#os.makedirs) to create directories recursively?

Example, when linting `Model/Payroll.php`: `/var/folders/5q/8gk0wq888xlggz008k8dr7180000hg/T/tmpE4plSY/Model/Payroll.php`

This will help linters that have rules applied based on regular expressions matching the basename *or* pathname. I guarantee these rules are almost always specified as a relative path anyway (say, `Model/*.php`, or `Controllers/*Controller.php`), which you would then have entirely reproduced in your temporary directory.

While this might be a bit more expensive than use just a filename, I wouldn't expect it to be a performance issue, and this would greatly improve backward compatibility with old linters until they get a chance to catch up with SL4.

As I said, I know zero about Python, so unfortunately I can not code a PR (your heard would explode).

**Comment by FichteFoll** [comment]
This seems to be very linter-specific. `mypy` allows substituting any specific file *in the project tree* with the file at the specified path, allowing to lint a file in a project but using the contents of a temporary file. This is a way better approach to it than trying to mirror the original folder structure, because we don't know how the linter will treat it, whether the structure should be relative to the project folder or the root and whether the linter supports that at all.



Can you name one or all linters that you know which:

1. have project-specific settings,
1. have settings that can apply to files by file path patterns,
1. can lint files at specific locations (temp folder) *while* considering the above two points?

Basically, if you run

```sh
cd /path/to/my/project
linter /var/folders/5q/8gk0wq888xlggz008k8dr7180000hg/T/tmpE4plSY/Controller/PayrollController.php
# OR
linter /var/folders/5q/8gk0wq888xlggz008k8dr7180000hg/T/tmpE4plSY/Controller/PayrollController.php --config-file=/path/to/my/project/linter.conf
```

does it produce the expected results with file path pattern settings applied?

**Comment by kaste** [comment]
Note guys, #1073 is the place to discuss new features, this here tracks an actionable issue namely to fix a regression.

**Comment by sebastienbarre** [comment]
@FichteFoll 

1. maybe I'm misunderstanding your question, but I can not name any linter that does *not* allow you to have project specific settings (maybe the `php` executable itself if you use it as a linter?). I don't know of many software projects that use the *exact* same *combo* of linting configurations for that matter, even in my own company. Consider the very opinionated [prettier/prettier](https://github.com/prettier/prettier) code formatter, that I use as a linter as well, it still allows for 3 or 4 config settings, because it's impossible to force people to choose between tabs or spaces for example. But to name a few that I use: csslint, eslint, flow, phpcs, phpmd. All of them, thank goodness,  have per-project settings (that can be stored as `phpcs.xml`, or `phpmd.xml`, or `.flowconfig` at the project root).
2. again, most  of them -- there are many good reasons for that, some of them [I listed here](https://github.com/SublimeLinter/SublimeLinter/issues/1073#issue-302027281). If you develop with any MVC framework you will likely need this option.
3. you are very correct here, I tried, and that didn't work, so consider my original suggestions moot: this apparently will not work because the linter gets the temporary path, then try to find its per-project config file to get it settings from, and fails to find it, since, of course, we are no longer in the root folder.

Personally, I would strongly suggest to at least pass the real filename to the linters if the SL is set to only lint on "load/save".

And if I may ask: why was it working perfectly fine in SL3 from an end-user perspective, and no  longer in SL4?

**Comment by braver** [comment]
>  why was it working perfectly fine in SL3 from an end-user perspective

Except that it wasn't in a lot of cases (multiple linters working on and trying to remove the same temp file). Also SL4 uses multi-threading more extensively. 

I still don't see how this is actionable. @FichteFoll raises issues related to what this issue proposes: it will simply not work and solve nothing. @kaste we're not "discussing new features", we're saying this issue is not a good idea.

> Personally, I would strongly suggest to at least pass the real filename to the linters if the SL is set to only lint on "load/save".

This already happens IIRC. load/save should only work on actual files in their actual location.

**Comment by kaste** [comment]
So basically we should *not* fix the regression.

**Comment by braver** [comment]
That is my current position. I think background linting with temporary files doesn't gel with configurations that "have settings that can apply to files by file path patterns" (@FichteFoll's 2nd point) and you should load/save mode for that. Which is why we need #244. Not that I'm not sympathetic to the issue at hand. I think perhaps LSP has a better chance of linting projects with such configurations, because language servers can be set up to load entire projects in memory and process your modifications on top of that. Linters are cool, but also too simple for some cases.

**Comment by FichteFoll** [comment]
>This already happens IIRC. load/save should only work on actual files in their actual location.

Not true. 

https://github.com/SublimeLinter/SublimeLinter/blob/ed9cbd277d656260e5382e4ceabed71437bee237/lint/linter.py#L1133-L1139

**Comment by sebastienbarre** [comment]
@braver: 

> This already happens IIRC. load/save should only work on actual files in their actual location.

I'm happy to hear that, but this was not the case two days ago, when SL4 landed, right? I opened an issue for it. At least one linter, SublimeLinter-phpcs, [had to be fixed](https://github.com/SublimeLinter/SublimeLinter-phpcs/pull/26). But again, if this is now the case for *all* linters, this  is great, and appreciated.

For background linting, the solution I described is better than just using a temporary file, for only a few more lines of code. It *does* help linters that allow adjustment per path patterns, as long as the linter config file is known to the linter -- all the ones I know allow you to pass the path to the config file as a command-line parameter. That's something some of the Sublime Linters packages allow you to configure as well. For those that don't, then whereas you use temporary files or temporary dirs with relative paths in them won't matter. But for those that do, it *does* make a difference, for no great cost IMHO. I've updated my comments above, the scenario @FichteFoll describe *does* work for me. It wasn't when I tried earlier because, according to the release notes, `If you use the ${project} settings token you should change that to ${folder}.` (ugh).  So now that I've updated my SL settings to this:
```
        "phpcs": {
            "disable": false,
            "args": [],
            "excludes": [
                "*.html"
            ],
            "standard": "${folder}/phpcs.xml"
        },
```
then it's all good, the file `/var/folders/5q/8gk0wq888xlggz008k8dr7180000hg/T/tmpE4plSY/Controller/PayrollController.php`, in Sublime, lints just as `Controller/PayrollController.php` lints from  its project folder.

**Comment by kaste** [comment]
To fix the regression would have been faster than to read this thread, honestly!

**Comment by sebastienbarre** [comment]
@kaste for real? *you* asked me for feedback!

> @sebastienbarre Want to tackle this one?

**Comment by kaste** [comment]
Yeah, but I set this issue to 'actionable' and now it's like 'wont_fix'

**Comment by FichteFoll** [comment]
I think the initial "regression" was only about choosing the original file's basename as either the suffix or prefix for the temporary file, but not trying to rebuild the folder structure in a tempdir.

Either way, I do not see how this used to work in SL3.

**Comment by braver** [comment]
> To fix the regression would have been faster than to read this thread, honestly!

Yeah, that happens. Sometimes it's good to re-check our positions.

I think it's curious that load/save doesn't just operate on the actual file though. Isn't that a bug? @FichteFoll have an opinion on that? I guess none of us actually use that mode much.

**Comment by FichteFoll** [comment]
@braver it's not a regression at least, I think.

**Comment by sebastienbarre** [comment]
Folks. It's Monday and people are getting back to work and auto-upgraded to SL4. Issues are piling on right now, and many seem to be regression ones. 

I don't know what the SL4 audience was. Developers? End users? From an end-user experience, I just want to lint and be productive today :) I don't know what benefits SL4 brought to be honest, and I appreciate all the great work, but here was my experience as an *end-user*:

- my user settings were gone (had to read the doc, compare both old and new), breaking linting right away.
- the handy tools menu has been removed, giving me the impression that maybe SL was no longer installed correctly.
- SL4 could no longer find my linter binaries (SL3 could, had to add hard-coded paths to SL settings).
- my linters looked broken (SL4 was passing temporary paths instead of the real paths).
- I could not quite read the linters output anymore (SL4 was no longer rendering/parsing HTML entities from linters output, while SL3 apparently could).
- I could not quite *see* the linters error because `error_color` or `warning_color` were gone (not fixable unless I start using a Sublime dev build).

I think I forgot some stuff, but it wasn't smooth... 

Some of them were fixed quickly, and honestly I can live without the "background" lint mode.  I would not expect that people read the release notes though, so if I *may* suggest, before you get an angry mob:
- please do back-port the SL3 settings to SL4, it's not too hard and will go a long way.
- I think there is still an issue with finding the binaries, and which PATH is being used by SL4. I can't be the only one having this issue, based on the new ones I see.

Just my 2 cents.

**Comment by braver** [comment]
> it's not too hard

Well, you do it then.

**Comment by kaste** [comment]
Now, we're all upset and the team splits LOL

But what the hell is a "regression". We changed the behavior, and this issue was to re-implement the old behavior while keeping the whole think thread-aware.

**Comment by FichteFoll** [comment]
This is the most useful feedback I have read about this update so far, so thanks for that.
Disclaimer: I'm just a contributor and participate in discussions. @kaste and @braver are the main maintainers.

>I would not expect that people read the release notes though

We have to, at some point. This isn't enterprise software where you pay for extended support periods. The old code base was a nightmare to maintain and many bugs were fixed for SL4 with new features added (like the panel and proper hover popups, or linter-specific executable overrides for *all* linters).

If we can't expect users to read release notes (and I mean the messages that pop up in Package Control when a package is updated, then what could we do instead?

>please do back-port the SL3 settings to SL4 automatically, it's not too hard and will go a long way. At least the "user" key.

This is a sensible suggestion. There used to be a command to convert the old settings, but there were loads of edge cases and it ended up not doing good work most of the time. I do believe that this is the best vector to make the upgrade experience much more smooth.

>SL4 could no longer find my linter binaries

This is 99% of the time macOS biting us ([troubleshooting section](http://www.sublimelinter.com/en/latest/troubleshooting.html#on-macos)). I think there is some work that can be done to improve the experience, but it doesn't seem like any of us is running a machine with macOS, so testing will be hard. If I had more time, I could try working on a band-aid fix until the upstream issue is addressed, but I have so much else to do that I just can't afford it.

In short: Because of some macOS oddities, SL3 tried to fetch the login shell's environment manually, while SL4 instead relies on Sublime Text doing that by itself (which makes it globally available to all plugins), but that unfortunately only completes *after* plugins are initialized.

>I could not quite see the linters error because error_color or warning_color were gone (not fixable unless I start using a Sublime dev build, which I'm trying).

There is a fall back for this, but it only works on dev builds currently. I believe it would have been a better idea to wait until the next stable ST build, but we have no idea when that is going to happen. I also didn't think of this at the time.

**Comment by billinghamj** [comment]
> please do back-port the SL3 settings to SL4 automatically, it's not too hard and will go a long way. At least the "user" key.

I don't get the impression that this is realistically possible. The vast majority of the removed options simply do not have equivalents in the new config, and it appears as though SL has no control over the actual colour rendering - other than suggesting that it would like something approximately "redish"

**Comment by sebastienbarre** [comment]
> The vast majority of the removed options simply do not have equivalents in the new config

I'm not referring to the colors at this point, I'm referring to *re-establishing basic functionality*.

I understand it's hard to hear as a developer, but right now, **SL4 functionality is broken for 100% of its end-users**, because everybody's settings were wiped out during upgrade.

So it's only the first business day after release, and you [get to read](https://github.com/SublimeLinter/SublimeLinter/issues/1102): "Honestly this seems like a bit of a disaster", or "Agreed, this is catastrophic to the point that I've abandoned Sublime altogether". This is not good. 

For a middleware point of view, a major release implies: update your dependencies at your own risk, possible API breakage, etc. This applies to downstream linters such as SublimeLinter-phpcs for example, though I would argue that since they couldn't specify which SL version they needed anyway, it's possibly an issue already.

From an *end-user* point of view, a major release *that they are forced to* means: most likely new functionality, maybe better performance, likely bug fixes. This is going to be most people's expectations. *Complete* loss of functionality is *not* going to be one of them :) And that's on top of a lot of things that were removed (the tools menu, color settings, tooltips, etc). That's two hard pills to swallow :) I've been trying to sound the alarm since Saturday, but I assume more people are going to enter issues this week, and you have better things to do, right?

* At least the "user" key  should be lifted up, to bring back people's linters config.
* Fix people's "@disable" to "disable", so disabled linters are not re-enabled.
* Fix people's `${project}`  to `${folder}`.

That's about it IMHO. The rest they can find in the doc.

**Comment by billinghamj** [comment]
The primary issue really is that the default behaviour of Package Control is to automatically update to whatever new versions are pushed, even if there is a new major version which isn't ready for production use yet.

If it hadn't updated automatically and just left us all on v3, I think everyone would be very happy and would have no problem with waiting until v4 was in a more works-off-the-shelf state.

**Comment by billinghamj** [comment]
> I'm not referring to the colors at this point, I'm referring to re-establishing basic functionality.

The colours are the bits which _do_ have rough equivalents in the new config :P The rest does not.

**Comment by braver** [comment]
> > please do back-port the SL3 settings to SL4 automatically, it's not too hard and will go a long way. > > > At least the "user" key.
> I don't get the impression that this is realistically possible.

You are correct. It looks easy, but it is super hard. 

> But what the hell is a "regression". We changed the behavior, and this issue was to re-implement the old behavior while keeping the whole thing thread-aware.

The old behavior doesn’t make sense, as I’ve pointed out before here.SL3 had a lot of stuff like that, that was just super wonky but probably solved someone’s problem, while causing more for others. We’re moving away from that stuff, and it can’t be helped that it breaks some stuff. The total end result is still positive though.

> This is 99% of the time macOS biting us (troubleshooting section). I think there is some work that can be done to improve the experience, but it doesn't seem like any of us is running a machine with macOS, so testing will be hard.

I run macOS. It’s a problem, but it’s self-healing and there are ways to work around it. The way SL3 messed with your path wasn’t acceptable either. 

> The primary issue really is that the default behaviour of Package Control is to automatically update to whatever new versions are pushed

Which is why we told everyone how to opt out. If no one reads that stuff that’s too bad but we did a pretty good job of putting that information out there. Manual upgrades are not a good solution either, we need to move forward and move 140 plugins along with us.

> SL4 functionality is broken for 100% of its end-users

Now you’re just ranting.
